### ISSUE TITLE
Rule suggestion: disallow boolean coercion

### ISSUE THREAD

**Issue by OliverJAsh** [palantir/tslint#4495]
I would like a rule to disallow boolean coercion, e.g.:

``` ts
declare const maybeUser: { username: string } | undefined;
type MyObject = { foo: 1 };
declare const myObject: MyObject;

// maybeUser is coerced to a boolean, so this should error
maybeUser && myObject;
```

This is not the same as [`strict-boolean-expressions`](https://palantir.github.io/tslint/rules/strict-boolean-expressions/), but it is similar.

`strict-boolean-expressions` would disallow the following code:

``` ts
declare const myBoolean: boolean;
type MyObject = { foo: 1 };
declare const myObject: MyObject;

// No operands are coerced to a boolean, so this should not error
myBoolean && myObject;
```

This is a case I do want to allow, because none of the operands undergo boolean coercion—the last operand (an object in this case) is only used as the result of the expression.

This is a common pattern in React, e.g. `shouldRender && <MyComponent />`.

**Comment by JoshuaKGoldberg** [comment]
@OliverJAsh agreed with your use case (`strict-boolean-expressions` is a big one to enable!), but it seems like it'd be confusing to have two rules, where one is a rough superset of the other. How about making `strict-boolean-expressions` have options to run the more lenient way you're proposing?

**Comment by OliverJAsh** [comment]
Is this enabled by https://github.com/palantir/tslint/pull/4159?

**Comment by adidahiya** [comment]
@OliverJAsh yeah, it looks like it is. I'm trying to get that released asap
